src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp

```diff
@@ -248,28 +251,28 @@ std::pair<std::vector<NativeInteger>, uint32_t> ParameterGenerationBGVRNS::compu
         totalModSize += modSize * (numPrimes - 1);
 
         // Compute moduli.
-        moduliQ[1] = FirstPrime<NativeInteger>(modSize, cyclOrder);
+        moduliQ[1] = LastPrime<NativeInteger>(modSize, cyclOrder);
         if (scalTech == FLEXIBLEAUTOEXT) {
             while (moduliQ[1] == moduliQ[0] || moduliQ[1] == moduliQ[numPrimes] || moduliQ[1] == plainModulusInt) {
-                moduliQ[1] = NextPrime<NativeInteger>(moduliQ[1], cyclOrder);
+                moduliQ[1] = PreviousPrime<NativeInteger>(moduliQ[1], cyclOrder);
             }
 
             for (size_t i = 2; i < numPrimes; i++) {
-                moduliQ[i] = NextPrime<NativeInteger>(moduliQ[i - 1], cyclOrder);
+                moduliQ[i] = PreviousPrime<NativeInteger>(moduliQ[i - 1], cyclOrder);
                 while (moduliQ[i] == moduliQ[0] || moduliQ[i] == moduliQ[numPrimes] || moduliQ[i] == plainModulusInt) {
-                    moduliQ[i] = NextPrime<NativeInteger>(moduliQ[i], cyclOrder);
+                    moduliQ[i] = PreviousPrime<NativeInteger>(moduliQ[i], cyclOrder);
```

Comment:
I am concerned that this logic with PrevousPrime can go to smaller bit size than requested. This is theoretically possible with BGV. We should probably put a logic to make sure the bitwidth does not become smaller.

Reply:
Generally speaking, the moduli should not become smaller. We may need to revisit this logic.

Reply:
I added checks in the code here to ensure the requested number of bits matches the returned number of bits.